src/sciwrappers.c

```diff
@@ -696,23 +696,24 @@ gint sci_get_lexer(ScintillaObject *sci)
 }
 
 
-void sci_set_lexer(ScintillaObject *sci, guint lexer_id)
+void sci_set_lexer_name(ScintillaObject *sci, const gchar *lexer_name)
 {
 	gint old = sci_get_lexer(sci);
 
-	/* TODO, LexerNameFromID() is already deprecated */
-	const char *lexer_name = LexerNameFromID(lexer_id);
-	if (! lexer_name)
+	g_return_if_fail(lexer_name != NULL);
+
+	old = sci_get_lexer(sci);
+
+	ILexer5 *lexer = CreateLexer(lexer_name);
+	if (! lexer)
 	{
-		g_warning("Failed to find lexer for ID %u", lexer_id);
+		g_warning("Failed to find lexer for name %s", lexer_name);
```

Comment:
I was wondering also in https://github.com/geany/geany/pull/3616 if it would be possible to do something similar like we do for the TM mappings checks - when starting Geany, going through all the lexers we use and check if we manage to create them; if not, hard-crash with `g_error()`. 

This would make errors like https://github.com/geany/geany/pull/3616 impossible. But this depends on two things:
1. `CreateLexer()` shouldn't be too expensive to call, I haven't checked what exactly it does.
2. We'd need to iterate through all lexers. Right now all lexers have to be mentioned explicitly at 2 different places in switches in `highlighting.c` and we probably don't want to add one more. So we'd have to create a table containing all the lexers, possibly in a similar way like in `filetypes.c`. As a side-effect one could then maybe avoid the warning from https://github.com/geany/geany/pull/2398 (which is produced by clang by default).

Thoughts?

Reply:
I don't know how expansive it'd be indeed, so whether or not it's reasonable to do it, but we could check.
From a quick look, it'll perform a linear search for the lexer of that name, and call Create() on the module, which either calls the lexer's `LexerFactoryFunction` (calling lexer code, which normally would instantiate an instance of the relevant lexer), or instantiate a new `LexerSimple`. So… it seems to be in between cheap and expansive, but I didn't profile anything.

> 2\. We'd need to iterate through all lexers. Right now all lexers have to be mentioned explicitly at 2 different places in switches in `highlighting.c` and we probably don't want to add one more. So we'd have to create a table containing all the lexers, possibly in a similar way like in `filetypes.c`.

I was wondering whether, now that we use a name for the lexers, if we'd want to avoid hard-coding things in highlightingmappings.h and move that to the filetypes files à-la-SciTE?  E.g. maybe something like this:

```
[lexer]
# could also be e.g. in [settings] as lexer_name
name = cpp
# style[lexer style number] = style name for use in [styling]
styles[0] = default
styles[1] = comment
styles[2] = commentline
styles[3] = commentdoc
# …
# keywors, those could possibly use GetWordListDescription() or similar to have actual names
keywords[0] = primary
keywords[1] = secondary
keywords[2] = docComment
# …
```

or if we were to break compatibility, we could avoid adding a proxy and just use the style/keyword numbers directly instead of mapping them twice..

I don't particularly like the idea of having arbitrary numbers there, but in the end, is it worse than the internal tables?  And maybe we could use `NamedStyles()`/`NameOfStyle()` and `DescribeWordListSets()` if enough lexers we need support that.

Anyway, it might not be relevant as we're not loading all filetype files anyway.

But note that any list to check would have to *actually* be synchronized with what is used; the issue in #3616 would only be found if checking the ID we *used*, not the list of lexers we *had*, as it was missing from the `AddLexerModules()` call.

Reply:
>  going through all the lexers we use and check if we manage to create them;

Making startup slower, would have to be avoided if geany is only being run to send a filename over the socket.

Better if all lexers are tested by CI not internal to Geany.

Reply:
To avoid speculations regarding the speed of creation of all lexers, I created #3673 testing it (less than 2 ms on Raspberry Pi 4). So I don't think we have to worry about speed here.

Reply:
> I was wondering whether, now that we use a name for the lexers, if we'd want to avoid hard-coding things in highlightingmappings.h and move that to the filetypes files à-la-SciTE? E.g. maybe something like this:

That would of course be great. No "built-in" or "external" filetypes, just builtin lexers and builtin ctags parsers which are configured dynamically in config files.

The other things like the indexed style definitions would be possible too, it would just be a little too magic for ordinary users as they don't know what's at the given index and would have to check the lexer sources.

> But note that any list to check would have to actually be synchronized with what is used; the issue in https://github.com/geany/geany/pull/3616 would only be found if checking the ID we used, not the list of lexers we had, as it was missing from the AddLexerModules() call.

Yes, sure.